File: 0076. Minimum Window Substring.cpp
class Solution
{
public:
    string minWindow(string s, string t)
    {
        if (s.size() == 0 or t.size() == 0)
        {
            return "";
        }
        
        map<char, int> tm;
        
        for (char c : t)
        {
            ++tm[c];
        }
        
        map<char, int> sm;
        int l = 0;
        int r = 0;
        int cs = 0;
        string ans = "";
        bool isAns = false;
        
        while (r < s.size())
        {
            ++sm[s[r]];
            
            if (tm.find(s[r]) != tm.end() and tm[s[r]] == sm[s[r]])
            {
                ++cs;
            }
            
            while (l <= r and cs == tm.size())
            {
                if (not isAns or r - l + 1 < ans.size())
                {
                    isAns = true;
                    ans = s.substr(l, r - l + 1);
                }
                
                --sm[s[l]];
                
                if (tm.find(s[l]) != tm.end() and tm[s[l]] > sm[s[l]])
                {
                    --cs;
                }
                
                ++l;
            }
            
            ++r;
        }
        
        return ans;
    }
    
};

File: 0076. Minimum Window Substring.cs
public class Solution
{
    public string MinWindow(string s, string t)
    {
        var frequencies = new int[58];

        foreach (var c in t)
        {
            ++frequencies[c - 'A'];
        }

        var l = 0;
        var r = 0;
        var currentFrequencies = new int[58];
        string result = null;

        while (r < s.Length)
        {
            var current = s[r] - 'A';
            ++currentFrequencies[current];

            while (ContainsEnough(currentFrequencies, frequencies))
            {
                if (result == null || r - l + 1 < result.Length)
                {
                    result = s[l..(r + 1)];
                }

                --currentFrequencies[s[l] - 'A'];
                ++l;
            }

            ++r;
        }

        return result != null ? result : string.Empty;
    }

    private static bool ContainsEnough(int[] current, int[] target)
    {
        for (int i = 0; i < 58; ++i)
        {
            if (current[i] < target[i])
            {
                return false;
            }
        }

        return true;
    }
}